src/mscorlib/shared/System/Globalization/TextInfo.Unix.cs

```diff
@@ -64,6 +65,54 @@ private unsafe string ChangeCase(string s, bool toUpper)
             return result;
         }
 
+        internal unsafe void ChangeCase(ReadOnlySpan<char> source, Span<char> destination, bool toUpper)
+        {
+            Debug.Assert(!_invariantMode);
+            Debug.Assert(destination.Length >= source.Length);
+
+            if (source.IsEmpty)
+            {
+                return;
+            }
+            
+            fixed (char* pSource = &MemoryMarshal.GetReference(source))
+            {
+                fixed (char* pResult = &MemoryMarshal.GetReference(destination))
+                {
+#if CORECLR
+                    if (IsAsciiCasingSameAsInvariant)
+                    {
+                        int length = source.Length;
+                        char* a = pSource, b = pResult;
+                        if (toUpper)
+                        {
+                            while (length-- != 0 && *a < 0x80)
+                            {
+                                *b++ = ToUpperAsciiInvariant(*a++);
+                            }
+                        }
+                        else
+                        {
+                            while (length-- != 0 && *a < 0x80)
+                            {
+                                *b++ = ToLowerAsciiInvariant(*a++);
+                            }
+                        }
+
+                        if (length != 0)
+                        {
+                            ChangeCase(a, source.Length - length, b, destination.Length - length, toUpper);
+                        }
+                    }
+                    else
+#endif
+                    {
+                        ChangeCase(pSource, source.Length, pResult, destination.Length, toUpper);
```

Comment:
Was there a conclusion on whether it is safe to depend on 1:1 mapping between lower and upper case?

https://github.com/dotnet/corefx/issues/21395#issuecomment-353233435

Reply:
I know that the current String APIs make this assumption, but that assumption does not affect their public shape - if needed, we can fix them without changing their public signature. 

It is not true for the Span APIs. If there is no 1:1 mapping, they need to start returning number of characters copied to output buffer.

Reply:
I think this is a reasonable suggestion for span (and maybe in the future we can support it for strings too). 

@ahsonkhan could you discuss this with the design committee to update such APIs? I know we mentioned we are not supporting this today but this doesn't prevent us to start supporting it with the spans. 

cc @KrzysztofCwalina

Reply:
Good point. If we change our API to return the copied characters, though redundant atm, it will be flexible to possible future implementation changes.

To be clear, is the new proposal for the APIs as follows (including the Invariant versions)?
```C#
public static int ToLower(this ReadOnlySpan<char> source, Span<char> destination, CultureInfo culture);
public static int ToLowerInvariant(this ReadOnlySpan<char> source, Span<char> destination);
public static int ToUpper(this ReadOnlySpan<char> source, Span<char> destination, CultureInfo culture);
public static int ToUpperInvariant(this ReadOnlySpan<char> source, Span<char> destination);
```

Reply:
> To be clear, is the new proposal for the APIs as follows (including the Invariant versions)?

Yes, Invariant is a culture :-)

Reply:
Question, what we'll do if the destination span is not enough? we throw? or return -1?

Reply:
I prefer that we throw just like we do atm for destination.Length < source.Length since the assumption is that the length required can be statically known at compile time prior to calling the method.

However, if the 1:1 mapping does not hold in the future, then my assumption breaks and throwing may not be the best solution.

Reply:
do we throw only because of the length? if so, I then prefer not throw and return -1 in case not enough buffer.

Reply:
For these APIs, should we enable partial processing (and ability to restart), like we do with some of our Base64 APIs, and return OperationStatus with charactersConsumed and charactersWritten out parameters? Or is that too much complexity for such a low level API, especially since there is a 1:1 mapping atm?

For example:
```C#
public static OperationStatus ToLower(this ReadOnlySpan<char>, Span<char> destination, CultureInfo culture, out char charactersConsumed, out char charactersWritten);
```

Reply:
we may have a simple one which just returns -1 and we can have the overload that takes more parameters if we find a demand for that.